### PR TITLE
Optimize manage phase of daemonset controller

### PR DESCRIPTION
**Pull request by jetmuffin**
### Ⅰ. Describe what this PR does

Optimize manage function when synchronize daemonsets.  In manage phase, daemonset controller figures out which nodes should run daemonset pods but not yet running and  which should not run but current running one. And then upgrades limited number of replicas according to partition number configured in RollingUpdate field.

However, current implementation does not work when is-first-deploy label is added to the daemonset. We found that when the label `is-first-deploy` is set to true, daemonset controller creates all new replicas in a batch no matter what the partition is; And if this label is set to false, daemonset controller deletes all old replicas once we change the node affinity or toleration in pod template spec.

### Ⅲ. List the added test cases (unit test/integration test) if any, please explain if no tests are needed.

Some conformance tests for daemonset controller has been added in this PR.

### CODE REVIEW

File: test/e2e/framework/daemonset_util.go [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+			}
+			if podutil.IsPodAvailable(&pod, ds.Spec.MinReadySeconds, metav1.Now()) {
+				nodesToPodCount[pod.Spec.NodeName] += 1
+			}
+		}
+		Logf("Number of nodes with available pods: %d", len(nodesToPodCount))
+
+		// Ensure that exactly 1 pod is running on all nodes in nodeNames.
+		for _, nodeName := range nodeNames {
+			if nodesToPodCount[nodeName] != 1 {
```
Comment (hantmac) [inline test/e2e/framework/daemonset_util.go:155]: Should be `if nodesToPodCount[nodeName] > 1 ` ?

File: pkg/controller/daemonset/daemonset_controller.go [hunk 2]
```diff
@@ -596,24 +596,29 @@ func (dsc *ReconcileDaemonSet) manage(ds *appsv1alpha1.DaemonSet, hash string) (
 		return delay, fmt.Errorf("couldn't get nodeList %v", err)
 	}
 
-	var nodesNeedingDaemonPods, podsToDelete []string
+	var nodesNeedingDaemonPods, nodesDesireScheduled, podsToDelete []string
```
Comment (FillZpp) [inline pkg/controller/daemonset/daemonset_controller.go:599]: can this `nodesDesireScheduled` be an integer?

**Review by FillZpp — COMMENTED** [review]
can we remove the `daemonset.kruise.io/is-first-deployed` label definition?

File: test/e2e/framework/daemonset_util.go (hunk 1 above)
Comment (jetmuffin) [reply test/e2e/framework/daemonset_util.go]: Okay, I'll fix this.

File: pkg/controller/daemonset/daemonset_controller.go (hunk 2 above)
Comment (jetmuffin) [reply pkg/controller/daemonset/daemonset_controller.go]: The last two comments have been fixed in the most recent commit :)

**Review by FillZpp — APPROVED** [review]
LGTM

**Pull request merged**